File: scripts/reset.mjs
#!/usr/bin/env node
// Reset the Configuration Manager's on-disk state (the data/ directory) for a clean restart.
//
// Usage:
//   node scripts/reset.mjs                 # wipe everything (the whole data/ directory)
//   node scripts/reset.mjs users           # clear only the user directory (re-triggers admin bootstrap)
//   node scripts/reset.mjs changes audit   # clear specific stores, keep the rest
//
// Targets: all (default) | users | changes | instances | settings | audit | repo
//
// Note: stop the API server before running this. The server keeps state in memory and rewrites
// these files, so deleting while it runs can be undone on its next save.
import { rmSync, existsSync } from 'node:fs';
import { dirname, join, resolve } from 'node:path';
import { fileURLToPath } from 'node:url';

// data/ lives at the repo root, one level up from this script — resolved independently of cwd.
const dataDir = resolve(dirname(fileURLToPath(import.meta.url)), '..', 'data');

const FILES = {
  users: 'users.json',
  changes: 'changes.json',
  instances: 'instances.json',
  settings: 'settings.json',
  audit: 'audit.json',
  repo: 'config-repo',
};

const args = process.argv.slice(2).map((a) => a.toLowerCase());
const targets = args.length === 0 || args.includes('all') ? ['all'] : args;

const unknown = targets.filter((t) => t !== 'all' && !(t in FILES));
if (unknown.length) {
  console.error(`Unknown target(s): ${unknown.join(', ')}`);
  console.error(`Valid targets: all, ${Object.keys(FILES).join(', ')}`);
  process.exit(1);
}

if (!existsSync(dataDir)) {
  console.log(`Nothing to reset — no data directory at ${dataDir}`);
  process.exit(0);
}

function remove(path, label) {
  if (existsSync(path)) {
    rmSync(path, { recursive: true, force: true });
    console.log(`  removed ${label}`);
  } else {
    console.log(`  (skip) ${label} — not present`);
  }
}

console.log(`Resetting Configuration Manager data in ${dataDir}`);
if (targets.includes('all')) {
  remove(dataDir, 'data/ (everything)');
} else {
  for (const t of targets) remove(join(dataDir, FILES[t]), `data/${FILES[t]}`);
}
console.log('Done. Restart the API and reload the app — the first visitor becomes admin.');

File: configManager-v2/reset.py
"""Reset the app's data from the command line. Mirrors the original `npm run reset`.

Usage:
  python reset.py                 # wipe everything (data/)
  python reset.py users           # clear only the users (re-arms admin bootstrap)
  python reset.py changes audit   # clear one or more stores

Targets: users | instances | changes | settings | audit | repo | live | all
  - "repo" clears both the built-in store (repo.json) and the git repo (config-repo/).
  - Each cleared store is re-created (re-seeded) on the next app start.

NOTE: Streamlit keeps the Store in memory (@st.cache_resource), so this CLI only takes
effect once the app is restarted — stop it (Ctrl+C) and run `streamlit run streamlit_app.py`
again. A browser reload alone is not enough. To reset a *running* app without a restart,
use Admin → Reset data instead.
"""
import os
import shutil
import stat
import sys
from pathlib import Path

DATA = Path(__file__).resolve().parent / "data"
# Each target maps to the path(s) it removes, relative to data/.
FILES = {
    "users": ["users.json"],
    "instances": ["instances.json"],
    "changes": ["changes.json"],
    "settings": ["settings.json"],
    "audit": ["audit.json"],
    "live": ["live.json"],
    "repo": ["repo.json", "config-repo"],  # built-in store + git repo
}


def _force_rmtree(path: Path) -> None:
    """Remove a tree even when git leaves read-only pack files behind (Windows)."""
    def on_error(func, p, _exc):
        os.chmod(p, stat.S_IWRITE)
        func(p)
    shutil.rmtree(path, onerror=on_error)


def _remove(path: Path, label: str) -> None:
    if not path.exists():
        print(f"  (already absent) {label}")
        return
    if path.is_dir():
        _force_rmtree(path)
    else:
        path.unlink()
    print(f"  removed {label}")


def main(argv):
    targets = argv or ["all"]
    if not DATA.exists():
        print(f"Nothing to reset — no data directory at {DATA}")
        return

    print(f"Resetting Configuration Manager data in {DATA}")
    if "all" in targets:
        _force_rmtree(DATA)
        print("  removed data/ (everything)")
    else:
        unknown = [t for t in targets if t not in FILES]
        if unknown:
            print(f"Unknown target(s): {', '.join(unknown)}")
            print(f"Valid targets: all, {', '.join(FILES)}")
            sys.exit(1)
        for t in targets:
            for rel in FILES[t]:
                _remove(DATA / rel, f"data/{rel}")
    print("Done. Restart the app (stop it and re-run streamlit) to pick up the reset.")


if __name__ == "__main__":
    main(sys.argv[1:])
